File: 55-jump-game/55-jump-game.scala
object Solution {
    def canJump(nums: Array[Int]): Boolean = {
        val l = nums.length - 1
        var f = l

        for(i <- 1 to nums.length - 1) {
            if (l - i + nums(l - i) >= f) {
                f = l - i
            }
        }

        f == 0
    }
}

File: 55-jump-game/55-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int maxReach = 0;
        
        for(int i = 0; i < nums.length; i++) {
            if (maxReach < i) return false;
            maxReach = Math.max(maxReach, i + nums[i]);
            if (maxReach >= nums.length - 1) return true;
        }
        
        return false;
    }
}